test/orchard/stacktrace_test.clj

```diff
@@ -0,0 +1,342 @@
+
+  (testing "Full file mappings"
+    (is (match?
+         (matchers/seq-of {:file-url #".+!/clojure/core.clj$"})
+         (filter #(= "clojure.core" (:ns %)) frames1)))
+    (is (match?
+         (matchers/seq-of {:file-url #"^file:/.+"})
+         (filter #(some->> (:ns %) (re-find #"^orchard")) frames1))))
+
+  (testing "Clojure ns, fn, and var"
+    ;; All Clojure frames should have non-nil :ns :fn and :var attributes.
+    (is (match? (matchers/seq-of {:ns some?, :fn some?, :var some?})
+                (filter #(= :clj (:type %)) frames1)))
+    (is (match? (matchers/seq-of {:ns some?, :fn some?, :var some?})
+                (filter #(= :clj (:type %)) frames2))))
+
+  (testing "Clojure name demunging"
+    ;; Clojure fn names should be free of munging characters.
+    (is (match? (matchers/seq-of {:fn (matchers/mismatch #"[_$]|(--\d+)")})
+                (filter :fn frames1)))
+    (is (match? (matchers/seq-of {:fn (matchers/mismatch #"[_$]|(--\d+)")})
+                (filter :fn frames2)))))
+
+(deftest stacktrace-frame-flags-test
+  (testing "Flags"
+    (testing "for file type"
+      ;; Every frame should have its file type added as a flag.
+      (is (every? #(contains? (:flags %) (:type %)) frames1))
+      (is (every? #(contains? (:flags %) (:type %)) frames2)))
+
+    (testing "for tooling"
+      ;; Tooling frames are classes named with 'clojure' or 'nrepl',
+      ;; or are java thread runners...or calls made from these.
+      (is (some #(re-find #"(clojure|nrepl|run)" (:name %))
+                (filter (comp :tooling :flags) frames1)))
+      (is (some #(re-find #"(clojure|nrepl|run)" (:name %))
+                (filter (comp :tooling :flags) frames2))))
+
+    (testing "for project"
+      (is (seq (filter (comp :project :flags) frames4))))
+
+    (testing "for duplicate frames"
+      ;; Index frames. For all frames flagged as :dup, the frame above it in
+      ;; the stack (index i - 1) should be substantially the same source info.
+      (let [ixd1 (zipmap (iterate inc 0) frames1)
+            ixd2 (zipmap (iterate inc 0) frames2)
+            dup? #(or (= (:name %1) (:name %2))
+                      (and (= (:file %1) (:file %2))
+                           (= (:line %1) (:line %2))))]
+        (is (every? (fn [[i v]] (dup? v (get ixd1 (dec ^long i))))
+                    (filter (comp :dup :flags val) ixd1)))
+        (is (every? (fn [[i v]] (dup? v (get ixd2 (dec ^long i))))
+                    (filter (comp :dup :flags val) ixd2)))))))
+
+(deftest exception-causes-test
+  (testing "Exception cause unrolling"
+    (is (= 2 (count causes1)))
+    (is (= 1 (count causes2))))
+  (testing "Exception data"
+    ;; If ex-data is present, the cause should have a :data attribute.
+    (is (:data (first causes1)))
+    (is (not (:data (first causes2))))))
+
+(deftest compilation-errors-test
+  (is (match? #"Syntax error compiling at \(.*orchard/stacktrace_test\.clj:"
```

Comment:
I prefer to have all assertions in `testing`, so that the output is a bit more descriptive, but it's not a big deal.